run-tests.php

```diff
@@ -1645,11 +1645,43 @@ function run_all_tests_parallel(array $test_files, array $env, $redir_tested): v
     }
 }
 
+/**
+ * Calls fwrite and retries when network writes fail with errors such as "Resource temporarily unavailable"
+ *
+ * @param resource $stream the stream to fwrite to
+ * @param string $data
+ * @return int|false
+ */
+function safe_fwrite($stream, string $data)
+{
+    // safe_fwrite was tested by adding $message['unused'] = str_repeat('a', 20_000_000); in send_message()
+    // fwrites on tcp sockets can return false or less than strlen if the recipient is busy.
+    // (e.g. fwrite(): Send of 577 bytes failed with errno=35 Resource temporarily unavailable)
+    $bytes_written = 0;
+    while ($bytes_written < strlen($data)) {
+        $n = @fwrite($stream, substr($data, $bytes_written));
+        if ($n === false) {
+            $write_streams = [$stream];
+            $read_streams = [];
+            $except_streams = [];
+            /* Wait for up to 10 seconds for the stream to be ready to write again. */
+            stream_select($read_streams, $write_streams, $except_streams, 10);
+            $n = @fwrite($stream, substr($data, $bytes_written));
+            if ($n === false) {
```

Comment:
```suggestion
            if (!stream_select($read_streams, $write_streams, $except_streams, 10) || @fwrite($stream, substr($data, $bytes_written)) === false) {
```

Reply:
although there won't be any error if stream_select times out so might be better to split it. It's just a good convention not to call fwrite if select doesn't succeed IMO. Not a big deal though in this case...

Reply:
I tested out your proposed change locally - when the messages were 20 megabytes it got stuck as a result, because stream_select was returning 0 despite the subsequent fwrite succeeding before.

With the current PR implementation (ignoring stream_select result) without the proposed change, run-tests.php would proceed normally and not get stuck

```patch
--- a/run-tests.php
+++ b/run-tests.php
@@ -1664,10 +1664,13 @@ function safe_fwrite($stream, string $data)
             $write_streams = [$stream];
             $read_streams = [];
             $except_streams = [];
+            fwrite(STDERR, "Going to call stream_select\n");
+            $result = stream_select($read_streams, $write_streams, $except_streams, 10);
+            if (!$result) {
+                fwrite(STDERR, "stream_select failed with " . var_export($result, true) . "\n");
+            }
             /* Wait for up to 10 seconds for the stream to be ready to write again. */
-            stream_select($read_streams, $write_streams, $except_streams, 10);
-            $n = @fwrite($stream, substr($data, $bytes_written));
-            if ($n === false) {
+            if (!$result || @fwrite($stream, substr($data, $bytes_written)) === false) {
                 echo "ERROR: send_message() Failed to write chunk after stream_select: " . error_get_last()['message'] . "\n";
                 return false;
             }
@@ -1679,6 +1682,7 @@ function safe_fwrite($stream, string $data)
 
 function send_message($stream, array $message): void
 {
+    $message['unused'] = str_repeat('a', 20_000_000);
     $blocking = stream_get_meta_data($stream)["blocked"];
     stream_set_blocking($stream, true);
     safe_fwrite($stream, base64_encode(serialize($message)) . "\n");
```

```
Going to call stream_selectt test workers running]
Going to call stream_select
Going to call stream_select
PASS #[AllowDynamicProperties] cannot be applied to interface [Zend/tests/allow_dynamic_properties_on_interface.phpt] 
PASS #[AllowDynamicProperties] cannot be applied to trait [Zend/tests/allow_dynamic_properties_on_trait.phpt] 
PASS Alternative offset syntax should emit E_COMPILE_ERROR in const expression [Zend/tests/alternative_offset_syntax_compile_error_in_const_expr.phpt] 
PASS Alternative offset syntax should emit E_COMPILE_ERROR outside const expression [Zend/tests/alternative_offset_syntax_compile_error_outside_const_expr.phpt] 
PASS bitwise AND and strings [Zend/tests/and_001.phpt] 
PASS declare bare anonymous class [Zend/tests/anon/001.phpt] 
PASS declare anonymous class extending another [Zend/tests/anon/002.phpt] 
PASS reusing anonymous classes [Zend/tests/anon/003.phpt] 
PASS testing anonymous inheritance [Zend/tests/anon/004.phpt] 
PASS testing reusing anons that implement an interface [Zend/tests/anon/005.phpt] 
PASS testing anon classes inside namespaces [Zend/tests/anon/006.phpt] 
PASS testing anon classes in functions outside of classes in namespaces [Zend/tests/anon/007.phpt] 
PASS testing static access for methods and properties in anon classes [Zend/tests/anon/008.phpt] 
PASS testing traits in anon classes [Zend/tests/anon/009.phpt] 
PASS Trait binding after anon class [Zend/tests/anon/010.phpt] 
PASS Ensure proper inheritance with get_class(anon class instance) used via class_alias (see also bug #70106) [Zend/tests/anon/011.phpt] 
PASS Ensure correct unmangling of private property names for anonymous class instances [Zend/tests/anon/012.phpt] 
PASS closure binding to anonymous class [Zend/tests/anon/013.phpt] 
PASS anonymous class trait binding [Zend/tests/anon/014.phpt] 
PASS static variables in methods inherited from parent class [Zend/tests/anon/015.phpt] 
PASS static variables in methods inherited from parent class (can't cache objects) [Zend/tests/anon/016.phpt] 
PASS Generated names for anonymous classes [Zend/tests/anon_class_name.phpt] 
PASS Basic argument unpacking [Zend/tests/arg_unpack/basic.phpt] 
PASS Argument unpacking with by-ref arguments [Zend/tests/arg_unpack/by_ref.phpt] 
PASS Array must be separated if unpacking by reference [Zend/tests/arg_unpack/by_ref_separation.phpt] 
PASS Unpack arguments for dynamic call [Zend/tests/arg_unpack/dynamic.phpt] 
PASS Identical comparison of array with references [Zend/tests/array_with_refs_identical.phpt] 
PASS Basic arrow function functionality check [Zend/tests/arrow_functions/001.phpt] 
PASS Arrow functions implicit use must be throwing notices only upon actual use [Zend/tests/arrow_functions/002.phpt] 
PASS Variable-variables inside arrow functions [Zend/tests/arrow_functions/003.phpt] 
PASS Auto-globals in arrow functions [Zend/tests/arrow_functions/004.phpt] 
PASS Arrow function $this binding [Zend/tests/arrow_functions/005.phpt] 
PASS Arrow functions syntax variations [Zend/tests/arrow_functions/006.phpt] 
PASS Pretty printing for arrow functions [Zend/tests/arrow_functions/007.phpt] 
PASS Yield inside arrow functions [Zend/tests/arrow_functions/008.phpt] 
PASS Bug #70528 (assert() with instanceof adds apostrophes around class name) [Zend/tests/assert/bug70528.phpt] 
PASS test passing assertion [Zend/tests/assert/expect_001.phpt] 
PASS test failing assertion [Zend/tests/assert/expect_002.phpt] 
PASS test catching failed assertion [Zend/tests/assert/expect_003.phpt] 
PASS test providing reason (fail) [Zend/tests/assert/expect_004.phpt] 
PASS test providing reason (pass) [Zend/tests/assert/expect_005.phpt] 
PASS test looping assert (pass) [Zend/tests/assert/expect_006.phpt] 
PASS test compiled reason [Zend/tests/assert/expect_007.phpt] 
PASS test disabled expectations have no ill side effects [Zend/tests/assert/expect_008.phpt] 
PASS test stack trace is correct from failed exception in extended class [Zend/tests/assert/expect_009.phpt] 
PASS test stack trace is correct from failed exception in extended class (parent implementing constructor) [Zend/tests/assert/expect_010.phpt] 
PASS test overloaded __toString on custom exception [Zend/tests/assert/expect_011.phpt] 
PASS test enable/disable assertions at runtime [Zend/tests/assert/expect_012.phpt] 
PASS test failing assertion when disabled (with return value) [Zend/tests/assert/expect_013.phpt] 
PASS test failing assertion when disabled [Zend/tests/assert/expect_014.phpt] 
PASS AST pretty-peinter [Zend/tests/assert/expect_015.phpt] 
PASS test enable/disable assertions at runtime (assertions not completely disabled) [Zend/tests/assert/expect_016.phpt] 
PASS test enable/disable assertions at runtime (assertions completely disabled) [Zend/tests/assert/expect_017.phpt] 
PASS test assertions in namespace [Zend/tests/assert/expect_018.phpt] 
PASS test assertions in namespace (assertions completely disabled) [Zend/tests/assert/expect_019.phpt] 
PASS AST pretty-printer [Zend/tests/assert/expect_020.phpt] 
PASS Empty statement in assert() shouldn't segfault [Zend/tests/assert/expect_empty_stmt_bug.phpt] 
PASS Return value of assigning by-val function result by-reference [Zend/tests/assign_by_val_function_by_ref_return_value.phpt] 
PASS Coalesce assign (??=): Basic behavior [Zend/tests/assign_coalesce_001.phpt] 
Going to call stream_selectt test workers running]
Going to call stream_select
Going to call stream_select
stream_select failed with 0
ERROR: send_message() Failed to write chunk after stream_select: fwrite(): Send of 20128377 bytes failed with errno=11 Resource temporarily unavailable
ERROR: Worker 2 reported unexpected E_NOTICE: unserialize(): Error at offset 0 of 68283 bytes in /usr/local/tyson/php-src/run-tests.php on line 1746
```

Reply:
More importantly, I still need to set `$n =` in that suggestion, it's used when falling through

Reply:
> More importantly, I still need to set $n = in that suggestion, it's used when falling through

Never mind, that was the only reason for the bug - the reader of the sent message would see bad data passed to $unserialize because $n wasn't set for the retried fwrite, meaning segments of the sent message were repeated.